lib/utilities/package-cache.js

```diff
@@ -0,0 +1,480 @@
+        return false;
+      }
+    }
+
+    return true;
+  },
+
+  /**
+   * The `_install` method installs the contents of the manifest into the
+   * specified package cache.
+   *
+   * @method _install
+   * @param {String} label The label for the cache.
+   * @param {String} type The type of package cache.
+   */
+  _install: function(label, type) {
+    commands[type].invoke('install', { cwd: this.dirs[label] });
+
+    // Link in Ember CLI after installation.
+    if (this.options.linkEmberCLI === true && (type === 'npm' || type === 'yarn')) {
+      commands[type].invoke('link', 'ember-cli', { cwd: this.dirs[label] });
+    }
+  },
+
+  /**
+   * The `_upgrade` method guarantees that the contents of the manifest are
+   * allowed to drift in a SemVer compatible manner. It ensures that CI is
+   * always running against the latest versions of all dependencies.
+   *
+   * @method _upgrade
+   * @param {String} label The label for the cache.
+   * @param {String} type The type of package cache.
+   */
+  _upgrade: function(label, type) {
+    // Unlink Ember CLI before upgrade.
+    if (this.options.linkEmberCLI === true && (type === 'npm' || type === 'yarn')) {
+      commands[type].invoke('unlink', 'ember-cli', { cwd: this.dirs[label] });
+    }
+
+    // Only way to get repeatable behavior in npm: start over.
+    // We turn an `_upgrade` task into an `_install` task.
+    if (type === 'npm') {
+      fs.removeSync(path.join(this.dirs[label], translate(type, 'path')));
+      this._install(label, type);
+    } else {
+      commands[type].invoke(translate(type, 'upgrade'), { cwd: this.dirs[label] });
+
+      // Re-link Ember CLI after upgrade.
+      if (this.options.linkEmberCLI === true && (type === 'npm' || type === 'yarn')) {
+        commands[type].invoke('link', 'ember-cli', { cwd: this.dirs[label] });
+      }
+    }
+  },
+
+  // PUBLIC API BELOW
+
+  /**
+   * The `create` method adds a new package cache entry.
+   *
+   * @method create
+   * @param {String} label The label for the cache.
+   * @param {String} type The type of package cache.
+   * @param {String} manifest The contents of the manifest file for the cache.
+   * @return {String} The directory on disk which contains the cache.
+   */
+  create: function(label, type, manifest) {
+
+    // TODO: Lock out upgrade calls after the first time accessing the cache.
+
+    // Set up the default Ember CLI link.
+    // Easier to do this every time than try and lock it out.
+    if (this.options.linkEmberCLI === true && (type === 'npm' || type === 'yarn')) {
+      var emberCLIPath = path.resolve(__dirname, '../..');
+      commands[type].invoke('link', { cwd: emberCLIPath });
+    }
+
+    // Compare any existing manifest to the ideal per current blueprint.
+    var identical = this._checkManifest(label, type, manifest);
+
+    if (identical) {
+      // Use what we have, but opt in to SemVer drift.
```

Comment:
@ejsundstr You make a good point that there may need to be additional cache key material. Using a cache across Node.js versions blows up the first time you have a native extension. Using a cache across `npm` versions _probably_ won't break you, but might. I'm going to abstract those two pieces directly into `PackageCache` so that the user doesn't have to think about it.